README.md

```diff
@@ -1 +1,191 @@
+- };
++ }, []);
+
+...
+```
+
+> NOTE: Now, just re-render the product we are updating.
+
+But as the application gets more complex, we may find other challenges, for example, we have to create a `FancyFilter` component to update the color and filter the product list:
+
+_./src/app.tsx_
+
+```diff
+import React from "react";
+import {
+  Product as UnmemoizedProduct,
+  AddProduct,
++ FancyFilter,
+} from "@/components";
+const Product = React.memo(UnmemoizedProduct);
++ import { useColor } from "@/hooks";
+- import { getUpdated } from "@/helpers";
++ import { getUpdated, getFiltered } from "@/helpers";
+import { INITIAL_PRODUCTS } from "./constants";
+import classes from "./app.module.css";
+
+- const ProductList = () => {
++ const ProductList = ({ visibility }) => {  
+  const [products, setProducts] = React.useState(INITIAL_PRODUCTS);
+  const handleChange = React.useCallback((id) => {
+    setProducts((products) => getUpdated(id, products));
+  }, []);
++ const filtered = getFiltered(products, visibility);
+
+  return (
+    <>
+      <ul>
+-       {products.map((product) => (
++       {filtered.map((product) => (
+          <Product key={product.id} product={product} onChange={handleChange} />
+        ))}
+      </ul>
+      <AddProduct onAdd={setProducts} />
+    </>
+  );
+};
+
+export const App = () => {
++ const { color, onColorChange } = useColor();
++ const [visibility, setVisibility] = React.useState("all");
+
+  return (
+    <div className={classes.root}>
++     <FancyFilter
++       color={color}
++       onColorChange={onColorChange}
++       visibility={visibility}
++       onVisibilityChange={setVisibility}
++     />
+-     <ProductList />
++     <ProductList visibility={visibility} />
+    </div>
+  );
+};
+
+```
+
+The above fix still works but we are calling the `getFiltered` function even if we only changes the color.
+
+In most applications this may not be a problem, but if we need a good user experience on devices with fewer resources, for example on mobile device, we may want to avoid unnecessary calculations.
+
+> NOTE: we can simulate it using the `Performance` tab in the browser dev tools.
+
+In that case, we need to use `React.useMemo` to memoize the result of the `getFiltered` function that depends on the `products` and `visibility`:
+
+_./src/app.tsx_
+
+```diff
+...
+
+const ProductList = ({ visibility }) => {
+  const [products, setProducts] = React.useState(INITIAL_PRODUCTS);
+  const handleChange = React.useCallback((id) => {
+    setProducts((products) => getUpdated(id, products));
+  }, []);
+- const filtered = getFiltered(products, visibility);
++ const filtered = React.useMemo(
++   () => getFiltered(products, visibility),
++   [products, visibility]
++ );
+
+...
+
+```
+
+Depending on the device it may go faster or slower, but we have improved somewhat.
+
+## SolidJS
+
+In this example we have implemented the same as the previous one but using SolidJS.
```

Comment:
In this example we have implemented the same functionality as in the previous one but using SolidJS.